Create SystemLimit data model

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Simple SQL table to hold onto limits for features. It is easier to update these via django admin than putting them in settings and having to run cchq commands, so that is the main inspiration for this change. This just introduces the model.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Very safe. Just adding a new table.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create FeatureLimit SQL table

gherceg committed:
Formatting file

gherceg committed:
Add FeatureLimit to admin

gherceg committed:
Ignore model in dump and restore

gherceg commented:
Cal happened to give me drive by thoughts on this, and mentioned that moving feature limits from settings to a model makes it harder to track changes (very fair point). The django admin does track who made changes and when, but that does not include actual values of the limit. We could add the field audit decorator to this model if we are interested in tracking changes over time, which I'm tempted to do.

orangejenny commented:
> We could add the field audit decorator to this model if we are interested in tracking changes over time, which I'm tempted to do.

That sounds like a good idea.

orangejenny commented on corehq/project_limits/models.py:
```diff
@@ -80,3 +81,12 @@ class RateLimitedTwoFactorLog(models.Model):
     status = models.CharField(max_length=31, null=False)
+
+
+class FeatureLimit(models.Model):
```
I thought these would have a domain, too - I added @nospame as a reviewer because this sounds like the spec he just wrote. Am I getting two features confused? @gherceg can you give an example of a feature?

gherceg replied:
Thanks for highlighting this Jenny! I was not aware of this and Evan and I offlined about it. It sounds like we are implementing two different things, where Evan's model is defining usage limits for specific projects/subscription, and this one is intended for system wide limits. So I'll likely rename this model to `SystemLimit` to align with that difference.

As for expecting these to have a domain, that is interesting. The standard case should not require a domain field, but in the event of an emergency, it would be nice to define a limit for a specific domain (whether higher or lower). So I'm tempted to add a domain field that defaults to null (meaning it applies to all domains), and we should only consider adding a specific domain entry in extreme circumstances. What do you think?

orangejenny replied:
I'd lean towards keeping it simpler for now. I can envision the domain-specific limits you're describing, but let's walk before we start running.

gherceg committed:
Rename FeatureLimit to SystemLimit

gherceg committed:
Audit SystemLimit limit field

last_modified is not needed since it can be determined via audit events

gherceg committed:
Add an optional domain field

This field is really only intended for extreme circumstances when we
want to either increase or decrease a specific domain's limit

gherceg committed:
Add unique constraint on key and domain

gherceg commented:
Sorry for all of the new commits. This is ready for review now.

orangejenny commented on corehq/project_limits/admin.py:
```diff
@@ -27,3 +28,9 @@ class PillowThrottleDefinitionAdmin(admin.ModelAdmin):
+
+@admin.register(SystemLimit)
+class SystemLimitAdmin(admin.ModelAdmin):
+    list_display = ('key', 'limit')
```
I think it'd be useful to add domain here.

gherceg replied:
[c8dc267](https://github.com/dimagi/commcare-hq/pull/35675/commits/c8dc2676badd5a7c16465349ceabd069879e57e4)

orangejenny commented on corehq/project_limits/models.py:
```diff
@@ -80,3 +82,17 @@ class RateLimitedTwoFactorLog(models.Model):
+        constraints = [models.UniqueConstraint(fields=['key', 'domain'], name='unique_key_per_domain_constraint')]
+
+    def __str__(self):
+        return f"{self.key}: {self.limit}"
```
I think it'd be useful to add domain here.

gherceg replied:
[c8dc267](https://github.com/dimagi/commcare-hq/pull/35675/commits/c8dc2676badd5a7c16465349ceabd069879e57e4)

gherceg replied:
[19c4551](https://github.com/dimagi/commcare-hq/pull/35675/commits/19c45511e031a4a304a57c4d918989f0dbaa307b)

gherceg committed:
Include domain in str repr and admin

gherceg committed:
Fix bug with str representation
